test/performance/search/index_construction_benchmark.cpp

```diff
@@ -0,0 +1,130 @@
+// Copyright (c) 2006-2019, Knut Reinert & Freie Universität Berlin
+// Copyright (c) 2016-2019, Knut Reinert & MPI für molekulare Genetik
+// This file may be used, modified and/or redistributed under the terms of the 3-clause BSD-License
+// shipped with this file and also available at: https://github.com/seqan/seqan3/blob/master/LICENSE
+// -----------------------------------------------------------------------------------------------------
+
+// @author: Clemens Cords <clemens.cords@fu-berlin.de>
+
+#include <benchmark/benchmark.h>
+#include <cstring>
+#include <seqan3/alphabet/all.hpp>
+#include <seqan3/search/fm_index/all.hpp>
+#include <seqan3/test/performance/sequence_generator.hpp>
+#include <seqan3/range/view/to_char.hpp>
+#include <seqan3/range/view/char_to.hpp>
+
+using namespace seqan3;
+
+unsigned int const SEED = 1234;
+
+// ============================================================================
+// construct as many indices as possible
+// ============================================================================
+template<typename index_t, std::size_t const length>
+void construct_index(benchmark::State & state)
+{
+    using alphabet_t = typename index_t::char_type;
+
+    auto sequence = test::generate_sequence<alphabet_t>(length, 0, SEED);
+
+    for (auto _ : state)
+        index_t index{sequence};
+}
+
+// show effect of alphabet size and compare fm with bi-fm
+BENCHMARK_TEMPLATE(construct_index, fm_index<std::vector<dna4>>, 25000);
+BENCHMARK_TEMPLATE(construct_index, fm_index<std::vector<aa27>>, 25000);
+BENCHMARK_TEMPLATE(construct_index, fm_index<std::vector<phred63>>, 25000);
+
+BENCHMARK_TEMPLATE(construct_index, bi_fm_index<std::vector<dna4>>, 25000);
+BENCHMARK_TEMPLATE(construct_index, bi_fm_index<std::vector<aa27>>, 25000);
+BENCHMARK_TEMPLATE(construct_index, bi_fm_index<std::vector<phred63>>, 25000);
+
+#if __has_include(<seqan/index.h>)
+
+// ============================================================================
+// export generated sequences for seqan2
+// ============================================================================
+
+// dna sequences shared by seqan2 and 3 benchmarks
+std::map<std::size_t, std::vector<char>> dna_sequence_dict;
+
+std::vector<char> get_sequence(std::size_t length)
+{
+    if (dna_sequence_dict.find(length) == dna_sequence_dict.end())
+    {
+        // generate new
+        auto sequence = test::generate_sequence<dna4>(length, 0, SEED);
+        dna_sequence_dict.insert({length, sequence | view::to_char});
+    }
+
+    return dna_sequence_dict[length];
+}
+
+// ============================================================================
+// seqan3 index construction with dna4 only
+// ============================================================================
+bool formatting_help_has_been_printed = false;
+
+template<typename index_t, std::size_t const length>
+void construct_index_dna(benchmark::State & state)
```

Comment:
why this function ? isn't `construct_index` enough because it is already templetized by the alphabet type?

Reply:
I needed to make a special case for the seqan3 benchmark function to react differently when called with dna4 (the same type the seqan2 function is called). 
``construct_index<... dna4>`` calls 
```cpp
30: auto sequence = test::generate_sequence<alphabet_t>(length, 0, SEED);
``` 

while 
``construct_index_dna`` calls
```cpp
80: auto sequence_char = get_sequence(length);
81: std::vector<dna4> sequence = sequence_char | view::char_to<dna4>;
```

This initializes the memory cache in line 52 and if it is already initalized just retrieves it. This is necessary so both the seqan3 and seqan2 function use the exact same sequence. My other option was to compare the template type against seqan3::dna4 and call ``get_sequence`` only then which I figured would be even more messy since that options is completely superflous if ``SEQAN3_HAS_SEQAN2`` evaluates to false (c.f. below). This way it's in a section of the code that's only relevant for the seqan2 comparison

Alternative way (pseudo code):
```cpp
template<typename index_t, std::size_t const length>
void construct_index(benchmark::State & state)
{
    using alphabet_t = typename index_t::char_type;

    if (std::is_same<alphabet_t, seqan3::dna4>:value)
        // call construct_index_dna method only for seqan2 comparison
    else
        // call original method

  (...)
}
```

Reply:
I couldn't think of a cleaner way